Update BotState to allow it to be used on its own as well as middleware

garypretty opened the pull request:
The full background to this discussion and it's intention can be found in #341. 

This PR is intended to make a small amend to how BotState works within the C# SDK to bring it in line with the JS SDK and allow it to be used independently, as well as from within middleware.  This is likely to eventually be superseded by more significant changes to state management, but this change will unblock the ability for middleware to maintain it's own state if required.

This change maintains the existing model for ConversationState and UserState.

Example use....

**Middleware constructor**

```cs
_state = new BotState<MiddlewareState>(new MemoryStorage(),
                $"BotState:{typeof(BotState<MiddlewareState>).Namespace}.{typeof(BotState<MiddlewareState>).Name}",
                (context) => $"botstate/{context.Activity.ChannelId}/{context.Activity.Conversation.Id}/{typeof(BotState<MiddlewareState>).Namespace}.{typeof(BotState<MiddlewareState>).Name}");
```

**Middleware process request**

```cs
var currentState = await _state.Read(context);
// do something with your state
await _state.Write(context, currentState);
```

cc'ing @billba @drub0y due to discussion in existing issue.

drub0y commented:
I would like to see the `keySelector` issue we discussed in #341 for `ConversationState/UserState<T>` fixed in this PR too. 

I think using `BotState<T>` directly is an even more advanced scenario as you're in charge of the entire namespace at that point and saving. What I would expect is that most people want will multiple `T`s of `ConversationState/UserState<T>` maintained for them and don't want to have to figure out how to tie it to those particular scopes themselves or get in the middle of saving (yet).

billba commented:
@drub0y That's a separate feature. Obviously using BotState directly is a lower level thing, but let's make it work, and then we can add more power to [C|U]State.

drub0y commented:
Ok, I just see those as a bug vs this is literally opening up the API to new uses. If we just want two separate PRs that's fine and I can follow on with that one if you want.

billba commented:
I see this as "align current C# with current JS" and then we can go to "add new feature to both C# and JS".

garypretty commented:
@billba @drub0y I agree with both of you 👍 @drub0y I am already on with a PR to bring key selector functionality on the back of this change to achieve what you have laid out but also to abstract the complexity when using bot state on its own.

Hopefully we can get this small change in to get alignment and then I'll submit the PR over the weekend as a first attempt and we can amend / tweak / debate the final implementation 😊 sound good?

garypretty commented:
@billba @drub0y btw. The way I am implementing key selection may be way off but as I was here I thought I'd have a go. No worries if people don't like it and we scrap for something else.

drub0y commented:
It's only "off" in the sense that your example uses `ConversationState<MiddlewareState>` when it really should be using a scheme unique to itself to avoid stepping on or being stepped on by `ConversationState<T>` at any point.

garypretty commented:
@drub0y agreed. Ill update the example shortly to show this. Also I have a plan for handling this nicely in the follow up PR around keys

garypretty commented:
@cleemullins any chance we can get this PR merged? At least bringing us in line with the JS Sdk until we have landed on a more permanent solution which is being discussed in parallel. If would unblock me on a couple of bits, especially ahead of next week's back. Cheers.

garypretty commented:
@cleemullins @drub0y @billba I am going to close this PR so that we can focus on the wider one that @drub0y has raised. It aligns nicely with this one and addresses the concerns I have with the current implementation.

drub0y commented:
In lieu of the State Management DCR PR (#376) being accepted, I believe this is a **P0** to fix. There is no way to have multiple, disparate pieces of conversation state without this fix. Specifically the scenario that is completely broken is that middleware cannot have it's own state object separate from the bot. The API allows you to configure/use multiple instances of `ConversationState<T>`, but the last instance to write to the store will always "win".

cleemullins commented:
@garypretty Thanks Gary!
